### PR TITLE
Add hasADMaterialProperty* APIs

### PR DESCRIPTION
**Pull request by lindsayad**
Refs #14701

### CODE REVIEW

File: test/src/kernels/ADMatDiffusionTest.C [hunk 1]
```diff
@@ -35,6 +35,19 @@ ADMatDiffusionTest::ADMatDiffusionTest(const InputParameters & parameters)
...
+      !defaultADMaterialProperty<Real>(deducePropertyName("ad_mat_prop")))
+    mooseError("It should be impossible to get an AD property without erroring and simultaneously "
+               "be neither able to retrieve the property with 'hasADMaterialProperty' nor through "
+               "a default property");
+  if (!hasMaterialProperty<Real>("regular_mat_prop") &&
+      !defaultMaterialProperty<Real>(deducePropertyName("regular_mat_prop")))
+    mooseError(
+        "It should be impossible to get a regular property without erroring and simultaneously "
+        "be neither able to retrieve the property with 'hasMaterialProperty' nor through "
+        "a default property");
```
Comment (loganharbour) [inline test/src/kernels/ADMatDiffusionTest.C:50]: Good test 👍

**Review by loganharbour — DISMISSED** [review]

**Comment by lindsayad** [comment]
wtf is this pronghorn failure. @snschune ?

**Comment by snschune** [comment]
One of those spurious failures, small magnitude and small relative difference. I'd relax the test criteria or exclude the offending variable from the test. 

We can start with regoldin though.

**Comment by lindsayad** [comment]
Oh wow this never got merged :-(

**Comment by snschune** [comment]
@lindsayad let me know if you need help with the Pronghorn failure.

**Comment by lindsayad** [comment]
Looks like it didn’t fail this time :-) @loganharbour may I get a re-review please?

**Review by loganharbour — APPROVED** [review]

**Pull request merged**